### PR TITLE
Use system-ui alias for UI fonts (fixes #10144)

### LINKED ISSUE #10144: Use system-ui as font on Linux
- VSCode Version: 1.4.0
- OS Version: Ubuntu MATE 16.04

This commit hard-coded the Ubuntu font for the UI: https://github.com/Microsoft/vscode/commit/11a3b51973fbfb384cdb702edee47fee9422849f

While this might look more integrated for Ubuntu Unity users, everywhere else (GNOME, KDE, MATE with other setting or another distro, ...) it doesn't. Having `sans-serif` as an alias which I can change to a font I want is better IMHO.

The best solution probably is to use the same font that menus are using, e. g. the system GUI/Gtk/system font. Is there a way to access this, maybe similar to `-apple-system` or `BlinkMacSystemFont`?

### LINKED ISSUE #28619: Text disappeared after update to 1.13 [Bug]
- VSCode Version: 1.13
- OS Version: Windows 10.0.14393

Steps to Reproduce:

1. Updated to 1.13
2. Nothing else

screenshot_1

**Update**

Insider version is the same

**Update 2**

Downgrade to 12.2 fixed the problem

**Update 3**

Problem is with **font-family: system-ui** , how to fix it?

### PR DESCRIPTION
**Pull request by kdrag0n**
This PR fixes #10144.

Commit 45d93e93255f9cdaa37955fdb79b4e460252b45b applied this change in some areas, but it was reverted to fix #28619. The underlying cause of the regression was [Chromium bug 733219](https://bugs.chromium.org/p/chromium/issues/detail?id=733219), which has now been fixed, so this change should be safe to apply now.

The old font stacks have been kept with lower priorities to work around [Chromium bug 724393](https://bugs.chromium.org/p/chromium/issues/detail?id=724393).

### CODE REVIEW

**Comment by bpasero** [comment]
Adding a couple more eyes to review this. Especially @Tyriar as the one that merged https://github.com/microsoft/vscode/pull/25570 should have a saying in this too.

My understanding is that this still would only impact Linux, but not Windows and macOS?

**Comment by joaomoreno** [comment]
I've triggered a product build on this so we can test it: https://dev.azure.com/monacotools/Monaco/_build/results?buildId=74119&view=results

**Review by joaomoreno — APPROVED** [review]
Works fine on my Windows box.

**Comment by bpasero** [comment]
Thanks @joaomoreno. I can test on macOS, maybe Daniel could on Linux.

**Review by miguelsolorio — APPROVED** [review]
Tested this on macOS Catalina and didn't see any differences, even when overlaying the two:

Left: This branch, Right: Current Insiders
image

**Review by mjbvz — APPROVED** [review]

**Comment by bpasero** [comment]
Adding @alexdima given there is a change to `Default standalone editor font`

**Comment by bpasero** [comment]
@kdrag0n I am worried when I see reports about not using `system-ui` such as:
* https://infinnie.github.io/blog/2017/systemui.html
* https://github.com/twbs/bootstrap/pull/22377

What can you say about the claim it will regress for users with non-us locale?

**Comment by kdrag0n** [comment]
I just set the system language to Chinese on my Windows VM and this change actually appears to improve consistency. 

Before:
image
Notice how the English font used in VS Code doesn't match Explorer, and the line height is different from the Chinese font. The height mismatch is especially evident in the menu mnemonics at the top, but it also shows in the "Help" section.

After:
image
With `system-ui`, VS Code's font looks more like Explorer and the English text looks more like it's a part of the Chinese text rather than being shorter.

The `system-ui` font can also be seen in Chrome's context menus:
image

And Explorer:
image

**Comment by bpasero** [comment]
There was also this user on Windows (https://github.com/microsoft/vscode/pull/25570#issuecomment-302182205) that reported bad results when we added `system-ui` as first font to pick. I am worried that we could get more users on Windows seeing a font that is different so I suggest we scope this change to Linux only. Does that make sense?

**Comment by kdrag0n** [comment]
The problem with #25570 was the removal of all the other fonts, caused by [Chromium bug 724393](https://bugs.chromium.org/p/chromium/issues/detail?id=724393) which is marked as wontfix. That's why I kept the old font stack after `system-ui` — it was cited as the fix for that issue in #26912.

I'm fine with limiting this change to Linux since that's where it affects me, but it's nice to be consistent across platforms in case someone wants to change the system font on another OS. Even without a custom system font, using `system-ui` on Windows helps fix the wrong font being used for Latin characters if the language is set to Chinese, as I demonstrated above.

I suspect there are probably similar cases of this fixing incorrect fonts with other languages and possibly even on macOS, though I haven't tested that.

**Comment by bpasero** [comment]
Let's roll with it in insiders and gather more feedback. I am merging this in given prior approval by people in https://github.com/microsoft/vscode/pull/25570 which is essentially the same change.

**Comment by fireattack** [comment]
Unfortunately, that issue mentioned in #25570 (or similar) still exists on non-English systems.

Setup: Win7 x64 Simplified Chinese system.

image

All the words are in proper English fonts.

image

All the words are in MS Yahei (default Chinese font for Simp. Chinese Windows), which is not optimal for displaying English.

(I use English UI to demonstrate the issue better. The same issue exists on Chinese UI too as soon as there are English letters.)

In general, I don't think using system-ui on Windows is a good idea, for the reasons stated in https://infinnie.github.io/blog/2017/systemui.html.

**Comment by bpasero** [comment]
@kdrag0n ?

**Comment by kdrag0n** [comment]
The new font is different, but according to what I saw when testing on a Chinese Windows 10 system, the new font is more correct. The old font was different from the English UI font used in the rest of the system, which changed when I set the language to Chinese. Are you sure that this isn't the case on Windows 7?

**Comment by fireattack** [comment]
I believe what you described is exactly the case for both Win 10 or Win 7; if you use "system-ui", it will apply a Chinese font ([MS YaHei](https://en.wikipedia.org/wiki/Microsoft_YaHei)) to all the glyphs, which is *not* optimized for Latin letters.

Yes, it is the Windows default font for Chinese locale, but no one likes it and it shouldn't be used on anything but Chinese characters. Same goes to Japanese and other Asian locales. 

Just to make it clear, this is not just my personal opinion, more or less an industry consensus. Please check the details in https://infinnie.github.io/blog/2017/systemui.html and why GitHub and [Bootstrap](https://github.com/twbs/bootstrap/pull/22377) along with batch of other major websites removed "system-ui" from their font stack.

Yes, this is a Windows only issue. So by all means use "system-ui" on *nix, just like `-apple-system` on Mac. But not Windows. If we want a catch-all-rule (doesn't seem necessary since we can define font stack for each OS already), putting it like `"Segoe UI", system-ui`, in this order. This would optimize the Latin letters on Windows with Segoe UI, while the Chinese characters will still fall back to a proper Chinese font (without affecting other OS since they don't have Segoe UI installed).

At the very least, the Chinese font shouldn't never be used for the *English UI*. This is not just about look and feel, but functionality: these Latin letters in those Chinese fonts are often wider, which cause the text to overflow all over the places. Probably less an issue for VS Code, but it has been a long-standing issue for any programs that doesn't test with non-Western locales since Windows 9X.


Feel free to ask some i18n experts from MS team for their insight, if above is not convincing.

**Comment by bpasero** [comment]
If that is the consensus, maybe another PR could be made for Windows only. Happy to review.

**Comment by kdrag0n** [comment]
I checked Bootstrap's [current font stack](https://github.com/twbs/bootstrap/blob/master/scss/_variables.scss#L386) and it looks like they brought system-ui back in https://github.com/twbs/bootstrap/pull/30561.

I'm still not entirely convinced about the Windows Chinese font issue because I think that apps should respect the system conventions whenever possible, even if it might not be the best option in the app developer's opinion. Inconsistent fonts (and mismatching line heights for CJK fonts, which can be distracting when reading mnemonics in menus) can be quite annoying. For example, in the blog post linked above, the system-ui font is used in the Chrome UI:
image

It's clearly not the best font for that content, but it *is* consistent with the rest of the system.

As far as I can tell, there's no instances of text overflow in VS Code either when using Microsoft YaHei for the UI.

However, if this change is to be reverted for Windows, adding Segoe UI before system-ui would cause problems on Linux and macOS systems that happen to have Segoe UI installed — this already introduces some inconsistency for me with the old font stack, where most of the UI is using Ubuntu but some parts are using Segoe UI.

This is especially an issue when it comes to the components that have fonts set by TypeScript code rather than CSS, e.g. VCS commit message editor and Markdown preview, as it's harder to use different font stacks based on the OS in those cases.

I'm not sure what the best way to proceed is in this case, considering consistency with the rest of the system vs. how the font looks when rendering Latin text. @bpasero thoughts?

**Comment by fireattack** [comment]
>it looks like they brought system-ui back in twbs/bootstrap#30561.

I tested 4.5.x before posting, didn't notice they changed it later. Sorry! But their wording there (https://github.com/twbs/bootstrap/pull/30561#issuecomment-618774485) doesn't sound to me they actually tested that i18n issues.

> the system-ui font is used in the Chrome UI

Chrome must have changed that too. This is what Simp. Chinese Chrome in Simp. Chinese Windows looks like:

image

You can see the font for Latin letters on tab is clearly not MS Yahei.

I checked some UI elements and this is what they use:

image

>  cause problems  .. that happen to have Segoe UI installed 

I agree. If we can do so per OS there is no need for such workaround. 

I get your point of consistency and understand that's also important. But I think at the very least, the UI shouldn't be in a Chinese font **if a Western display language is chosen in VS Code**.

**Comment by bpasero** [comment]
@kdrag0n 

Can we not simply change the rule from:

`font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif;`

to 

`font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", system-ui, "Ubuntu", "Droid Sans", sans-serif;`

My understanding is that:
* `-apple-system, BlinkMacSystemFont` are ignored on Linux, Windows
* `"Segoe WPC", "Segoe UI"` are ignored on Linux, macOS
* `system-ui` will be picked up by Linux distros

Wouldn't this solve the original intent of fixing the font stack on Linux? It would not change a lot for Windows I guess, but given there seem to be issues, maybe we shouldn't try to solve it.

**Comment by kdrag0n** [comment]
> `"Segoe WPC", "Segoe UI"` are ignored on Linux, macOS

As I said in my previous comment, this *can* cause issues on other OSes:

> Adding Segoe UI before system-ui would cause problems on Linux and macOS systems that happen to have Segoe UI installed — this already introduces some inconsistency for me with the old font stack, where most of the UI is using Ubuntu but some parts are using Segoe UI.

---

What do you think about the fix in https://github.com/kdrag0n/vscode/commit/91ac1372c4c0e8bce6ff3d2ab50d635a44b5524c? I haven't tested it myself, but it's similar to what @fireattack suggested above and it appears to be a reasonable compromise while avoiding issues on other setups. Let me know if you want me to create a new PR for it.

**Comment by bpasero** [comment]
@kdrag0n sorry I am confused now, my suggestion was to add `system-ui` AFTER `Segoe` not before. In essence this should really only have an impact on Linux.

Yes, the fix you linked to seems going into the right direction, however this PR changed the font in many more places, some of them have no notion of OS. Hence my suggestion to fix it for all OS given the suggestion I made.

**Comment by kdrag0n** [comment]
The problem with putting system-ui after Segoe UI is that Linux or macOS systems that have Segoe UI installed. While that font is typically only found on Windows systems, it is still possible to install on other OSes and this can be useful in many cases, e.g. creating mockups.  

I believe that changes would need to be made to the TS code and the CSS selectors for the other places would need to be made in order to fix the issue for all possible setups.

**Comment by bpasero** [comment]
@kdrag0n I understand, but isn't the presence of the Ubuntu font on a non-Ubuntu distro a good indication that maybe that font should be used?

**Comment by kdrag0n** [comment]
Not really. Many Linux users (me included) aren't using Ubuntu and have the Ubuntu font installed for one reason or another, but that doesn't mean it should be used. On Linux, the UI font is customizable and there are many people who do change it, so VS Code should respect the user's choice. It's one of the only apps on my system that ignores my preferences and chooses to use a different font. That's the reason I opened this PR in the first place: to fix the font disparity and use the correct font. It's hard to generalize OSes and use a single font stack for all of them because of the Windows CJK issue, so I think the only option is to make the other places OS-dependent as well.

I'm not entirely sure whether this is the case, but one plausible out-of-the-box example is Kubuntu where the Ubuntu font would be installed because it's a Ubuntu-based distro, but the UI font would be set to Noto Sans as per KDE's defaults.

**Comment by bpasero** [comment]
@kdrag0n given we do not have platform specific rules in some places, why can we not do this rule instead, putting `system-ui` after `Segoe`:

```
-apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif
```

My understanding:
* macOS: is not impacted, `-apple-system, BlinkMacSystemFont` takes care of that
* Windows: is not impacted, `Segoe` is winning
* Linux: should only be impacted if `Segoe` is actually installed, but will prefer a native Linux font over `Ubuntu`

So this would only cause issues if someone installed `Segoe` intentionally, but even in that case today already VSCode would pick this font up.

**Comment by kdrag0n** [comment]
Yes, that would work fine for most setups, except Linux systems that have Segoe installed. That's one of the problems that I originally intended to solve with this PR as I'm seeing both inconsistent (Segoe) and non-native (Ubuntu) UI fonts with the old font stacks. If your proposed umbrella stack is used in other places, the Segoe inconsistency issue would come back and some Linux users would see inconsistent fonts again — the very issue that this was supposed to fix.

Again, I don't see any way to fix this issue everywhere unless the other places are updated to add platform-specific stacks (or unify all/most of the scattered stacks into one somehow to reduce maintenance costs) because system UI fonts are an inherently platform-specific detail.

**Comment by bpasero** [comment]
@kdrag0n upon closer looking, it seems to me we are already inconsistent given that our platform rules for the workbench define a different font stack compared to those places where we have the combined font stack:

https://github.com/microsoft/vscode/blob/af3d37848c49e763416cf927dcaf65ff3091cfd7/src/vs/workbench/browser/media/style.css#L8-L24

Maybe I can get the platform differences into the other places too.

Btw on Windows we already have different fonts if your locale is `zh-Hans`, `zh-Hant`, `ja` or `ko`. I think you mentioned that for those, picking `system-ui` is a better fit over `Segoe`, so we could leave `system-ui` in front for those at least.

**Comment by fireattack** [comment]
> for those, picking system-ui is a better fit over Segoe

The opposite. Having `Segoe` *in front* of `system-ui` in these locales can ensure the Latin letters are in `Segoe` while the CJK characters still fall back to system fonts because Segoe doesn't have these glyphs.

Actually, as you can see here, since corresponding system font is already in the rule ("Microsoft YaHei" for zh-Hans for example), there is no need for "system-ui" at all.

Same goes to Western locales: the `system-ui` is already `Segoe`.

To summarize: system-ui isn't needed on Windows at all, and putting it in front of `Segoe` causes harm on `zh`/`ja/`ko` locale (which is exactly why we had things like `"Segoe WPC", "Segoe UI", "Microsoft YaHei"` (in this order) before.

**Comment by bpasero** [comment]
@kdrag0n @fireattack I tried my best in https://github.com/microsoft/vscode/pull/99429 please take a look.

**Comment by fireattack** [comment]
Looks good to me, will test once it launched on Insider.

**Comment by bpasero** [comment]
@fireattack we have since released a new insiders with these changes

**Comment by fireattack** [comment]
Just tested, it looks good in both Simplified Chinese and English on my Simplified Chinese Windows.

image

image

Thanks!

**Comment by bpasero** [comment]
Thanks for taking the time to test this 👍
